### PR TITLE
Refactor functests to have re-useable fixtures for being logged in

### PR DESCRIPTION
**Pull request by jon-betts**
This first changes some fixtures into our situational style "with_logged_in_user" and then moves them into a file so they can be used by other tests.

### CODE REVIEW

File: tests/functional/fixtures/authentication.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+    login_user()
+
+
+@pytest.fixture
+def login_user(db_session, app, user):
+    def login_user():
+        # This is the hash for `pass` used below
+        user.password = "$2b$12$21I1LjTlGJmLXzTDrQA8gusckjHEMepTmLY5WN3Kx8hSaqEEKj9V6"
+        db_session.commit()
```
Comment (jon-betts) [inline tests/functional/fixtures/authentication.py:40]: This commit means any changes made to the user prior to this call take effect, like the `user.admin = True` above as well as setting the password.

The original had the password set on creation, but I think I prefer this version as it keeps the setting of the hash and using the value in the same place. Previously it was a bit obscure as to why this was being set.

File: tests/functional/fixtures/authentication.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
+import pytest
+
+__all__ = (
+    "user",
+    "with_logged_in_user",
+    "with_logged_in_staff_member",
+    "with_logged_in_admin",
+)
```
Comment (jon-betts) [inline tests/functional/fixtures/authentication.py:9]: An all here prevents us from re-exporting `pytest` which looks odd in other places that import things.

File: tests/functional/fixtures/authentication.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+def user(factories):
+    return factories.User()
+
+
+@pytest.fixture
+def login_user(db_session, app, user):
+    def login_user():
+        # This is the hash for `pass` used below
+        user.password = "$2b$12$21I1LjTlGJmLXzTDrQA8gusckjHEMepTmLY5WN3Kx8hSaqEEKj9V6"
+        db_session.commit()
```
Comment (jon-betts) [inline tests/functional/fixtures/authentication.py:22]: The commit here means the other fixtures which do things like `user.staff = True` will be persisted to the DB. It also keeps the hash and the value of the hash together. Previously it was a bit unclear why this was being set.

File: tests/functional/conftest.py [hunk 4]
```diff
@@ -7,6 +7,7 @@
 from tests.common.fixtures import es_client  # noqa: F401
 from tests.common.fixtures import init_elasticsearch  # noqa: F401
 from tests.common.fixtures.elasticsearch import ELASTICSEARCH_INDEX, ELASTICSEARCH_URL
+from tests.functional.fixtures.authentication import *  # noqa: F403,F401
```
Comment (seanh) [inline tests/functional/conftest.py:10]: I kind of prefer just keeping all the global and shared fixtures in the `conftest.py` files, to be honest. It's a more obvious location, makes them easier to find. There shouldn't be too many of them anyway. Over-organizing just makes the code more of a pain to navigate

**Review by seanh — APPROVED** [review]
:+1: Definitely a good general set of fixtures to have

**Pull request merged**